src/algebra/group/prod.lean

```diff
@@ -0,0 +1,236 @@
+  .. prod.monoid, .. prod.has_inv }
+
+instance [add_comm_semigroup A] [add_comm_semigroup B] : add_comm_semigroup (A × B) :=
+{ add_comm := assume a b, mk.inj_iff.mpr ⟨add_comm _ _, add_comm _ _⟩,
+  .. prod.add_semigroup }
+@[to_additive add_comm_semigroup]
+instance [comm_semigroup G] [comm_semigroup H] : comm_semigroup (G × H) :=
+{ mul_comm := assume a b, mk.inj_iff.mpr ⟨mul_comm _ _, mul_comm _ _⟩,
+  .. prod.semigroup }
+
+instance [add_comm_monoid A] [add_comm_monoid B] : add_comm_monoid (A × B) :=
+{ .. prod.add_comm_semigroup, .. prod.add_monoid }
+@[to_additive add_comm_monoid]
+instance [comm_monoid M] [comm_monoid N] : comm_monoid (M × N) :=
+{ .. prod.comm_semigroup, .. prod.monoid }
+
+instance [add_comm_group A] [add_comm_group B] : add_comm_group (A × B) :=
+{ .. prod.add_comm_semigroup, .. prod.add_group }
+@[to_additive add_comm_group]
+instance [comm_group G] [comm_group H] : comm_group (G × H) :=
+{ .. prod.comm_semigroup, .. prod.group }
+
+end prod
+
+namespace monoid_hom
+
+variables (M N) [monoid M] [monoid N]
+
+/-- Given monoids `M`, `N`, the natural projection homomorphism from `M × N` to `M`.-/
+@[to_additive "Given additive monoids `A`, `B`, the natural projection homomorphism
+from `A × B` to `A`"]
+def fst : M × N →* M := ⟨prod.fst, rfl, λ _ _, rfl⟩
+
+/-- Given monoids `M`, `N`, the natural projection homomorphism from `M × N` to `N`.-/
+@[to_additive "Given additive monoids `A`, `B`, the natural projection homomorphism
+from `A × B` to `B`"]
+def snd : M × N →* N := ⟨prod.snd, rfl, λ _ _, rfl⟩
+
+/-- Given monoids `M`, `N`, the natural inclusion homomorphism from `M` to `M × N`. -/
+@[to_additive "Given additive monoids `A`, `B`, the natural inclusion homomorphism
+from `A` to `A × B`."]
+def inl : M →* M × N :=
+⟨λ x, (x, 1), rfl, λ _ _, prod.ext rfl (one_mul 1).symm⟩
+
+/-- Given monoids `M`, `N`, the natural inclusion homomorphism from `N` to `M × N`. -/
+@[to_additive "Given additive monoids `A`, `B`, the natural inclusion homomorphism
+from `B` to `A × B`."]
+def inr : N →* M × N :=
+⟨λ y, (1, y), rfl, λ _ _, prod.ext (one_mul 1).symm rfl⟩
+
+variables {M N}
+
+@[to_additive] lemma coe_fst : ⇑(fst M N) = prod.fst := rfl
+@[to_additive] lemma coe_snd : ⇑(snd M N) = prod.snd := rfl
+
+@[to_additive] lemma inl_apply (x) : inl M N x = (x, 1) := rfl
+@[to_additive] lemma inr_apply (y) : inr M N y = (1, y) := rfl
+
+@[simp, to_additive] lemma fst_comp_inl : (fst M N).comp (inl M N) = id M := rfl
+@[simp, to_additive] lemma snd_comp_inl : (snd M N).comp (inl M N) = 1 := rfl
+@[simp, to_additive] lemma fst_comp_inr : (fst M N).comp (inr M N) = 1 := rfl
+@[simp, to_additive] lemma snd_comp_inr : (snd M N).comp (inr M N) = id N := rfl
+
+@[simp, to_additive] lemma fst_inl (x) : (inl M N x).1 = x := rfl
+@[simp, to_additive] lemma snd_inl (x) : (inl M N x).2 = 1 := rfl
+@[simp, to_additive] lemma fst_inr (y) : (inr M N y).1 = 1 := rfl
+@[simp, to_additive] lemma snd_inr (y) : (inr M N y).2 = y := rfl
```

Comment:
```suggestion
```

These would be made redundant by `inl_apply` and `inr_apply`. I don't want to have to use extensionality for pairs before `[simp]` lemmas will fire.